library/mps/reader.c

```diff
@@ -0,0 +1,626 @@
+                 * resuming on a reader than what we tried before pausing it.
+                 * If we believe we adhere to this restricted usage throughout
+                 * the library, this check is a good opportunity to
+                 * validate this. */
+                RETURN( MBEDTLS_ERR_READER_INCONSISTENT_REQUESTS );
+            }
+        }
+
+        acc += end;
+        *buffer = acc;
+        if( buflen != NULL )
+            *buflen = desired;
+
+        end += desired;
+        rd->end = end;
+        rd->pause = 0;
+
+        RETURN( 0 );
+    }
+
+    TRACE( trace_comment, "Serve the request from the current fragment." );
+
+    /* Attempt to serve the request from the current fragment */
+    fl = rd->frag_len;
+    frag_fetched = end - fo;
+    frag += frag_fetched;
+    frag_remaining = fl - frag_fetched;
+
+    if( frag_remaining < desired )
+    {
+        TRACE( trace_comment, "There's not enough data in the current fragment to serve the request." );
+        /* There's not enough data in the current fragment,
+         * so either just RETURN what we have or fail. */
+        if( buflen == NULL )
+        {
+            rd->pause = desired - frag_remaining;
+            TRACE( trace_comment, "Remember to collect %u bytes before re-opening",
+                   (unsigned) rd->pause );
+            RETURN( MBEDTLS_ERR_READER_OUT_OF_DATA );
+        }
+
+        desired = frag_remaining;
+    }
+
+    *buffer = frag;
+    if( buflen != NULL )
+        *buflen = desired;
+
+    end += desired;
+    rd->end = end;
+    rd->pause = 0;
+    RETURN( 0 );
+}
+
+int mbedtls_reader_commit( mbedtls_reader *rd )
+{
+    unsigned char *frag, *acc;
+    size_t aa, end, fo, shift;
+    TRACE_INIT( "reader_commit" );
+
+    /* Check that the reader is in consuming mode. */
+    frag = rd->frag;
+    if( frag == NULL )
+        RETURN( MBEDTLS_ERR_READER_UNEXPECTED_OPERATION );
+
+    acc = rd->acc;
+    end = rd->end;
+
+    if( acc == NULL )
+    {
+        TRACE( trace_comment, "No accumulator, just shift end" );
+        rd->commit = end;
+        RETURN( 0 );
+    }
+
+    fo = rd->acc_share.frag_offset;
+    if( end >= fo )
+    {
+        TRACE( trace_comment, "Started to serve fragment, get rid of accumulator" );
+        shift = fo;
+        aa = 0;
+    }
+    else
+    {
+        TRACE( trace_comment, "Still serving from accumulator" );
+        aa = rd->acc_avail;
+        shift = end;
+        memmove( acc, acc + shift, aa - shift );
+        aa -= shift;
+    }
+
+    end -= shift;
+    fo -= shift;
+
+    rd->acc_share.frag_offset = fo;
+    rd->acc_avail = aa;
+    rd->commit = end;
+    rd->end = end;
+
+    TRACE( trace_comment, "Final state: (end=commit,fo,avail) = (%u,%u,%u)",
+           (unsigned) end, (unsigned) fo, (unsigned) aa );
+    RETURN( 0 );
+}
+
+int mbedtls_reader_reclaim( mbedtls_reader *rd, size_t *paused )
+{
+    unsigned char *frag, *acc;
+    size_t pause, commit;
```

Comment:
I would rename "pause" to "pending", or maybe "missing", both here and in the mbedtls_reader. "pause", without any indication that it is a length, is very misleading.

Reply:
Agreed, a name indicating more clearly that it's an amount of data rather than a boolean flag would be better here.